File: elixir/allergies/lib/allergies.ex
defmodule Allergies do
  @allergens [:eggs, :peanuts, :shellfish, :strawberries, :tomatoes, :chocolate, :pollen, :cats]

  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags) do
    @allergens
    |> Stream.zip(little_endian_bits(flags))
    |> Stream.filter(fn {_, bit} -> bit == 1 end)
    |> Stream.map(fn {allergen, _} -> "#{allergen}" end)
    |> Enum.to_list()
  end

  defp little_endian_bits(number) do
    number
    |> Integer.digits(2)
    |> Enum.reverse()
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item), do: item in list(flags)
end

File: ocaml/allergies/allergies.ml
type allergen =
  | Eggs
  | Peanuts
  | Shellfish
  | Strawberries
  | Tomatoes
  | Chocolate
  | Pollen
  | Cats

let allergen_exp = function
  | Eggs -> 0
  | Peanuts -> 1
  | Shellfish -> 2
  | Strawberries -> 3
  | Tomatoes -> 4
  | Chocolate -> 5
  | Pollen -> 6
  | Cats -> 7

let allergen_score allergen =
  let exp = allergen_exp allergen in
  let rec aux acc exp = if exp = 0 then acc else aux (2 * acc) (exp - 1) in
  aux 1 exp

let allergic_to score allergen = score land allergen_score allergen <> 0

let allergies score =
  let allergens =
    [
      Eggs; Peanuts; Shellfish; Strawberries; Tomatoes; Chocolate; Pollen; Cats;
    ]
  in
  List.fold_right
    (fun a acc -> if allergic_to score a then a :: acc else acc)
    allergens []
